polytope_server/common/request_store/dynamodb_request_store.py

```diff
@@ -0,0 +1,243 @@
+            self.metric_store = metric_store.create_metric_store(metric_store_config)
+
+        logger.info("DynamoDB request store configured for table name %s.", table_name)
+
+    def get_type(self):
+        return "dynamodb"
+
+    def add_request(self, request):
+        try:
+            self.table.put_item(Item=_dump(request), ConditionExpression=Attr("id").not_exists())
+        except botocore.exceptions.ClientError as e:
+            if e.response["Error"]["Code"] == "ConditionalCheckFailedException":
+                raise ValueError("Request already exists in request store") from e
+            raise
+
+        if self.metric_store:
+            self.metric_store.add_metric(RequestStatusChange(request_id=request.id, status=request.status))
+
+        logger.info("Request ID {} status set to {}.".format(request.id, request.status))
+
+    def remove_request(self, id):
+        try:
+            self.table.delete_item(Key={"id": id}, ConditionExpression=Attr("id").exists())
+        except botocore.exceptions.ClientError as e:
+            if e.response["Error"]["Code"] == "ConditionalCheckFailedException":
+                raise KeyError("Request does not exist in request store") from e
+            raise
+
+        if self.metric_store:
+            items = self.metric_store.get_metrics(request_id=id)
+            for item in items:
+                self.metric_store.remove_metric(item.uuid)
+
+        logger.info("Request ID %s removed.", id)
+
+    def get_request(self, id):
+        response = self.table.get_item(Key={"id": id})
+        if "Item" in response:
+            return _load(response["Item"])
+
+    def get_requests(self, ascending=None, descending=None, limit=None, status=None, user=None, **kwargs):
+        if ascending is not None and descending is not None:
+            raise ValueError("Cannot sort by ascending and descending at the same time.")
+
+        query = _make_query(**kwargs)
+        if user is not None:
+            key_cond_expr = Key("user_id").eq(str(user.id))
+            fn = self.table.query
+            params = {
+                "IndexName": "user-index",
+                "KeyConditionExpression": key_cond_expr,
+            }
+            if status is not None:
+                query["status"] = status.value
+        elif status is not None:
+            key_cond_expr = Key("status").eq(status.value)
+            fn = self.table.query
+            params = {
+                "IndexName": "status-index",
+                "KeyConditionExpression": key_cond_expr,
+            }
+        else:
+            fn = self.table.scan
+            params = {}
+
+        if query:
+            filter_expr = reduce(operator.__and__, (Attr(key).eq(value) for key, value in query.items()))
+            params["FilterExpression"] = filter_expr
+
+        if limit is not None:
+            params["Limit"] = limit
+
+        reqs = (_load(item) for item in _iter_items(fn, **params))
+        if ascending:
```

Comment:
Same as before, perhaps there is a way to delegate the sorting to the db. `ScanIndexForward`?